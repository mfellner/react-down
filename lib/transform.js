import React from 'react'
import MarkdownIt from 'markdown-it'
import camelCase from 'camel-case'

function parseStyle(string) {
  const style = {}
  const pairs = string.split(';')
  for (let pair of pairs) {
    const [key, value] = pair.split(':')
    style[camelCase(key)] = value
  }
  return style
}

function parseAttrs(attrs) {
  if (!Array.isArray(attrs)) return null
  else return attrs.reduce((props, attr) => {
      const key = camelCase(attr[0])
      const value = key === 'style' ? parseStyle(attr[1]) : attr[1]
      return Object.assign(props, {[key]: value})
  }, {})
}

function createReactElements(nodes, tokens) {
  const lastNode = () => nodes[nodes.length - 1]
  let key = 0

  for (let token of tokens) {
    key += 1
    switch(token.type) {
      case 'heading_open':
      case 'paragraph_open':
      case 'em_open':
      case 'strong_open':
      case 'link_open':
      case 'blockquote_open':
      case 'bullet_list_open':
      case 'ordered_list_open':
      case 'list_item_open':
      case 'table_open':
      case 'thead_open':
      case 'tbody_open':
      case 'th_open':
      case 'tr_open':
      case 'td_open':
      case 's_open':
        nodes.push({
          type: token.tag,
          props: parseAttrs(token.attrs),
          children: []
        })
        break
      case 'inline':
        createReactElements(nodes, token.children)
        break
      case 'text':
        lastNode().children.push(token.content)
        break
      case 'softbreak':
      case 'hr':
        lastNode().children.push(React.createElement(token.tag, {key}))
        break
      case 'code_block':
      case 'code_inline':
      case 'fence':
        lastNode().children.push(React.createElement(token.tag, {key}, token.content))
        break
      case 'image':
      {
        const props = Object.assign({key}, parseAttrs(token.attrs), {alt: token.content})
        lastNode().children.push(React.createElement(token.tag, props))
        break
      }
      case 'heading_close':
      case 'paragraph_close':
      case 'em_close':
      case 'strong_close':
      case 'link_close':
      case 'blockquote_close':
      case 'bullet_list_close':
      case 'ordered_list_close':
      case 'list_item_close':
      case 'table_close':
      case 'thead_close':
      case 'tbody_close':
      case 'th_close':
      case 'tr_close':
      case 'td_close':
      case 's_close':
      {
        const node = nodes.pop()
        const props = Object.assign({key}, node.props)
        const element = React.createElement(node.type, props, node.children.slice())
        lastNode().children.push(element)
        break
      }
      default:
        console.warn('Unknown token:', token)
    }
  }
}

function createReactTree(tokens, type = 'div') {
  const root = {
    type,
    props: null,
    children: []
  }
  createReactElements([root], tokens)
  return React.createElement(root.type, root.props, root.children)
}

function parseMarkdown(src) {
  const md = new MarkdownIt()
  return md.parse(src, {})
}

export default function transform(src) {
  return createReactTree(parseMarkdown(src))
}
