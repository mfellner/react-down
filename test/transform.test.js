import React from 'react'
import renderer from 'react-test-renderer'
import transform from '../lib/transform'

describe('transform', () => {

  it('should transform headings', () => {
    const src = '# Hello, world!\n## This is a test.'

    const component = renderer.create(transform(src))
    expect(component.toJSON()).toMatchSnapshot()
  })

  it('should transform paragraphs', () => {
    const src = 'First.\n\nSecond.'

    const component = renderer.create(transform(src))
    expect(component.toJSON()).toMatchSnapshot()
  })

  it('should transform emphasis', () => {
    const src = '*Italics.* **Bold!**'

    const component = renderer.create(transform(src))
    expect(component.toJSON()).toMatchSnapshot()
  })

  it('should transform links', () => {
    const src = '[Link](https://github.com)'

    const component = renderer.create(transform(src))
    expect(component.toJSON()).toMatchSnapshot()
  })

  it('should transform images', () => {
    const src = '![test](test.png)'

    const component = renderer.create(transform(src))
    expect(component.toJSON()).toMatchSnapshot()
  })

  it('should transform blockquotes', () => {
    const src = '> some quote'

    const component = renderer.create(transform(src))
    expect(component.toJSON()).toMatchSnapshot()
  })

  it('should inline code', () => {
    const src = 'This is `code!`'

    const component = renderer.create(transform(src))
    expect(component.toJSON()).toMatchSnapshot()
  })

  it('should transform horizontal rules', () => {
    const src = '----'

    const component = renderer.create(transform(src))
    expect(component.toJSON()).toMatchSnapshot()
  })

  it('should transform tables', () => {
    const src = '| Col A | Col B |\n'
    + '| ------:| -----------:|\n'
    + '| aaa    | bbb         |\n'

    const component = renderer.create(transform(src))
    expect(component.toJSON()).toMatchSnapshot()
  })
})
