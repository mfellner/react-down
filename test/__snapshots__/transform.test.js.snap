exports[`transform should inline code 1`] = `
<div>
  <p>
    This is 
    <code>
      code!
    </code>
  </p>
</div>
`;

exports[`transform should transform blockquotes 1`] = `
<div>
  <blockquote>
    <p>
      some quote
    </p>
  </blockquote>
</div>
`;

exports[`transform should transform emphasis 1`] = `
<div>
  <p>
    <em>
      Italics.
    </em>
     
    <strong>
      Bold!
    </strong>
  </p>
</div>
`;

exports[`transform should transform headings 1`] = `
<div>
  <h1>
    Hello, world!
  </h1>
  <h2>
    This is a test.
  </h2>
</div>
`;

exports[`transform should transform horizontal rules 1`] = `
<div>
  <hr />
</div>
`;

exports[`transform should transform images 1`] = `
<div>
  <p>
    <img
      alt="test"
      src="test.png" />
  </p>
</div>
`;

exports[`transform should transform links 1`] = `
<div>
  <p>
    <a
      href="https://github.com">
      Link
    </a>
  </p>
</div>
`;

exports[`transform should transform paragraphs 1`] = `
<div>
  <p>
    First.
  </p>
  <p>
    Second.
  </p>
</div>
`;

exports[`transform should transform tables 1`] = `
<div>
  <table>
    <thead>
      <tr>
        <th
          style={
            Object {
              "textAlign": "right",
            }
          }>
          Col A
        </th>
        <th
          style={
            Object {
              "textAlign": "right",
            }
          }>
          Col B
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td
          style={
            Object {
              "textAlign": "right",
            }
          }>
          aaa
        </td>
        <td
          style={
            Object {
              "textAlign": "right",
            }
          }>
          bbb
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
